### PR TITLE
unsnap: busy wait a number of cycles 

### PR DESCRIPTION
**Pull request by vuule**
Busy replacement for `nanosleep`.

### CODE REVIEW

File: cpp/src/io/comp/unsnap.cu [hunk 1]
```diff
@@ -29,6 +29,16 @@ constexpr int32_t batch_count   = (1 << 2);
 constexpr int32_t prefetch_size = (1 << 9);  // 512B, in 32B chunks
 constexpr bool log_cyclecount   = false;
 
+void __device__ busy_wait(size_t cycles)
+{
+  clock_t start = clock();
+  for (;;) {
+    clock_t const now    = clock();
+    clock_t const passed = now > start ? now - start : now + (0xffffffff - start);
+    if (passed >= cycles) return;
```
Comment (harrism) [inline cpp/src/io/comp/unsnap.cu:38]: I think "passed" is overloaded, so "elapsed" is clearer.

```suggestion
    clock_t const elapsed = now > start ? now - start : now + (0xffffffff - start);
    if (elapsed >= cycles) return;
```

File: cpp/src/io/comp/unsnap.cu [hunk 2]
```diff
@@ -280,7 +291,7 @@ __device__ void snappy_decode_symbols(unsnap_state_s *s, uint32_t t)
     if (t == 0) {
       s->q.prefetch_rdpos = cur;
 #pragma unroll(1)  // We don't want unrolling here
-      while (s->q.prefetch_wrpos < min(cur + 5 * batch_size, end)) {}
+      while (s->q.prefetch_wrpos < min(cur + 5 * batch_size, end)) { busy_wait(10); }
```
Comment (harrism) [inline cpp/src/io/comp/unsnap.cu:294]: Why 10 in some places and 20 in others?

**Review by harrism — APPROVED** [review]
Minor suggestions. Otherwise looks fine.

File: cpp/src/io/comp/unsnap.cu (hunk 2 above)
Comment (vuule) [reply cpp/src/io/comp/unsnap.cu:294]: The original times in nanoseconds were 50 or 100. I noticed a significant perf regression when using the same number of cycles (compared to **no** waits). Divided all wait cycles by 5 and got great perf results. Went with this to make the PR available for testing ASAP.
In short, the numbers are not carefully chosen, I'm open to suggestions on the values to use here.

**Comment by abellina** [comment]
With this change, the hang reported https://github.com/rapidsai/cudf/issues/8055 is not seen anymore. We've tested small batches of queries, and last night triggered a much wider run and didn't see issues. So far so good @vuule.

**Review by elstehle — APPROVED** [review]

**Review by devavret — APPROVED** [review]

**Pull request merged**